src/main/java/org/quiltmc/loader/impl/config/Json5Serializer.java

```diff
@@ -0,0 +1,248 @@
+
+	private Json5Serializer() {
+
+	}
+
+	@Override
+	public String getFileExtension() {
+		return "json5";
+	}
+
+	private void serialize(JsonWriter writer, Object value) throws IOException {
+		if (value instanceof Integer) {
+			writer.value((Integer) value);
+		} else if (value instanceof Long) {
+			writer.value((Long) value);
+		} else if (value instanceof Float) {
+			writer.value((Float) value);
+		} else if (value instanceof Double) {
+			writer.value((Double) value);
+		} else if (value instanceof Boolean) {
+			writer.value((Boolean) value);
+		} else if (value instanceof String) {
+			writer.value((String) value);
+		} else if (value instanceof ValueList<?>) {
+			writer.beginArray();
+
+			for (Object v : (ValueList<?>) value) {
+				serialize(writer, v);
+			}
+
+			writer.endArray();
+		} else if (value instanceof ValueMap<?>) {
+			writer.beginObject();
+
+			for (Map.Entry<String, ?> entry : (ValueMap<?>) value) {
+				writer.name(entry.getKey());
+				serialize(writer, entry.getValue());
+			}
+
+			writer.endObject();
+		} else if (value instanceof ConfigSerializableObject) {
+			serialize(writer, ((ConfigSerializableObject<?>) value).getRepresentation());
+		} else if (value == null) {
+			writer.nullValue();
+		} else if (value.getClass().isEnum()) {
+			writer.value(((Enum<?>) value).name());
+		} else {
+			throw new ConfigParseException();
+		}
+	}
+
+	private void serialize(JsonWriter writer, ValueTreeNode node) throws IOException {
+		for (String comment : node.metadata(Comment.TYPE)) {
+			writer.comment(comment);
+		}
+
+		if (node instanceof ValueTreeNode.Section) {
+			writer.name(node.key().getLastComponent());
+			writer.beginObject();
+
+			for (ValueTreeNode child : ((ValueTreeNode.Section) node)) {
+				serialize(writer, child);
+			}
+
+			writer.endObject();
+		} else {
+			TrackedValue<?> trackedValue = ((TrackedValue<?>) node);
+			Object defaultValue = trackedValue.getDefaultValue();
+
+			if (defaultValue.getClass().isEnum()) {
+				StringBuilder options = new StringBuilder("options: ");
+				Object[] enumConstants = defaultValue.getClass().getEnumConstants();
+
+				for (int i = 0, enumConstantsLength = enumConstants.length; i < enumConstantsLength; i++) {
+					Object o = enumConstants[i];
+
+					options.append(o);
+
+					if (i < enumConstantsLength - 1) {
+						options.append(", ");
+					}
+				}
+
+				writer.comment(options.toString());
+			}
+
+			for (Constraint<?> constraint : trackedValue.constraints()) {
+				writer.comment(constraint.getRepresentation());
+			}
+
+			if (!(defaultValue instanceof CompoundConfigValue<?>)) {
+				writer.comment("default: " + defaultValue);
+			}
+
+			writer.name(node.key().getLastComponent());
+
+			serialize(writer, trackedValue.getRealValue());
+		}
+	}
+
+	@Override
+	public void serialize(Config config, OutputStream to) throws IOException {
+		JsonWriter writer = JsonWriter.json5(new OutputStreamWriter(to));
```

Comment:
Allowing strict json would be essentially free here (the only major concern would be to not write comments in strict mode), so it is probably worth implementing a JsonSerializer too?

Reply:
Even if it is free I really don't see a point to doing so.